File: Stack/simplifyPath.py
'''
0071 Simplify Path

stack
'''

class Solution:
    def simplifyPath(self, path: str) -> str:
        # split the folder names by '/'
        folders = path.split('/')
        # use a stack to store those folder names
        stack = []

        for folder in folders:
            # '..' correspond to the parent folder!
            if folder == '..':
                if stack: # if the stack if none empty, pop out the parent folder
                    stack.pop()
            elif folder and folder != '.':
                # otherwise, current folder has a name & not refer to current folder
                # append it to the stack
                stack.append(folder)
        # remember to add the root '/'
        return '/' + '/'.join(stack)

File: Stack/simplifyPath.cpp
/**
0071 Simplify Path
Cpp version
*/

class Solution 
{
public:
    string simplifyPath(string path) 
    {
        auto split = [](const string& s, char delim) -> vector<string> 
        {
            vector<string> ret;
            string cur;
            for (char c : s) 
            {
                if (c == delim) 
                {
                    ret.push_back(move(cur));
                    cur.clear();
                }
                else 
                {
                    cur += c;
                }
            }
            ret.push_back(move(cur));
            return ret;
        };
        vector<string> dirs = split(path, '/');
        vector<string> myStack;
        for (string& dir: dirs) 
        {
            if (dir == "..") 
            {
                if (!myStack.empty()) 
                {
                    myStack.pop_back();
                }
            }
            else if (!dir.empty() && dir != ".") 
            {
                myStack.push_back(move(dir));
            }
        }
        string result;
        if (myStack.empty()) 
        {
            result = "/";
        }
        else 
        {
            for (string& dir: myStack) 
            {
                result += "/" + move(dir);
            }
        }
        return result;
    }
};